src/man/peochk.8
.\"	$CoreSDI: peochk.8,v 1.8 2000/06/14 23:22:18 claudio Exp $
.\"
.\" Copyright (c) 2000
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"    This product includes software developed by Core-SDI SA and its
.\"    contributors.
.\" 4. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 10, 2000
.Dt PEOCHK 8
.Os Core-SDI
.Sh NAME
.Nm peochk
.Nd Initial key generator and integrity log file checker
.Sh SYNOPSIS
.Nm peochk
.Op Fl f Ar logfile
.Op Fl g
.Op Fl h
.Op Fl i Ar key0file
.Op Fl k Ar keyfile
.Op Fl l
.Op Fl m Ar hash_method
.Op Fl q
.Op Ar logfile
.Sh DESCRIPTION
.Nm peochk
generates the initial key file and checks log files generated by
.Xr syslogd 8
using
.Em peo output module
.Xr om_peo 8 .
The options are as follows:
.Bl -tag -width Ds
.It Fl f Ar logfile
Specify the pathname of a log file, if
.Ar logfile
is not specified using this option, data is readed from standard input
and the pathname is used only to generate reports and/or to obtain the
key files pathnames when the
.Fl k
and/or
.Fl i
options are not specified; the default is
.Pa /var/log/messages .
.It Fl g
Generates two key files with an initial key into them, one in binary mode
(
.Ar keyfile ,
to be used by
.Em peo output module
) and the other in ascii mode (
.Ar key0file
), the admin should put the
last one into a secure place and remove it from the specified path (see 
.Fl i
and 
.Fl k
options); when this option is not specified
.Nm
is in check mode.
.It Fl h
Displays a little help.
.It Fl i Ar key0file
Specify the initial key pathname; the default is
.Ar keyfile
pathname with a "0" char added at the end (see 
.Fl k
option).
.It Fl k Ar keyfile
Specify the key pathname (this file is used by the
.Em peo output module
to generate a hash key from the last logged message); the default is
.Pa /var/ssyslogd/xxx.key
where 
.Pa xxx
is
.Ar logfile
(specified with 
.Fl f
option or without it) with all '/' replaced by '.'.
.It Fl l
Used only in check mode to detect the first corrupted line; it is ignored
when specified with the
.Fl g
option.
.It Fl m Ar hash_method
Specifies the hash method used to generate the keys,
.Ar hash_method
should be one of 
.Cm md5, sha1,
or
.Cm rmd160;
the default is
.Cm sha1.
.It Fl q
Quiet mode; prints '0' on stdout when logfile is not corrupted, and '1' or
line number (see 
.Fl l
option) when the logfile is corrupted.
.El
.Sh EXAMPLES
If you want to protect the
.Pa /var/log/authlog
file you can:
.Pp
.Bl -enum
.It
run the command:
.Pp
.Dl peochk -g -f /var/log/authlog -i authkey0 -m rmd160
.Pp
this will generate the
.Pa /var/ssylog/var.log.authlog.key
file with the initial key in binary mode and the
.Ar ./authkey0
file with that key translated to ascii, the hash method used to generate
the key is
.Cm rmd160;
you should memorice the contents of
.Ar ./authkey0
file and
.Xr rm 1
it.
.Pp
.It
Edit
.Xr syslogd.conf 5
file and enable
.Em peo output module
with something like this:
.Pp
.Dl auth.info  %classic /var/log/authlog %peo -m rmd160 -l -k /var/ssyslog/.var.log.authlog.key
.Pp
.It
Inform new changes on
.Xr syslogd.conf 5
to
.Xr syslogd 8 :
.Pp
.Dl kill -HUP `cat /var/run/syslog.pid`
.Pp
.It
When you believe that someone owned your machine you can:
.Pp
.Dl peochk -m rmd160 -f /var/log/authlog -i mykey
.Pp
the contents of
.Ar mykey
should be the same as 
.Ar ./authkey0
generated in step 1; with the command above you can verify that the
file was (or not) corrupted (it is important not to forget the 
.Fl m
option because the default used is 
.Cm sha1
and the keys generated was using 
.Cm rmd160
).
.El
.Sh SEE ALSO
.Xr om_peo 8 ,
.Xr syslogd 8 ,
.Xr syslog.conf 5
.Sh BUGS
if you found bugs, please report them to syslog-bugs@core-sdi.com


NEWS
/*	$CoreSDI$	*/

Release Notes:

* Move to autoconf 
* Reorganize the source tree
* Fix an incorrect free that was causing syslogd to coredump on exit


src/peo/peochk.c
/*      $Id: peochk.c,v 1.14 2000/05/04 00:19:14 claudio Exp $
 *
 * peochk - syslog -- Initial key generator and integrity log file checker
 *
 * Author: Claudio Castiglia for Core-SDI SA
 *
 *
 * peochk [-f logfile] [-g] [-i key0file] [-k keyfile] [-l] [-m hash_method] [logfile]
 *
 * supported hash_method values:
 *			md5
 *			rmd160
 *			sha1
 *
 * defaults:
 *	logfile: 	/var/log/messages
 *	keyfile:	/var/ssyslog/.var.log.messages.key
 *	hash_method:	sha1
 *
 * NOTES:
 *	1) When logfile is specified without the -f switch, the data is
 *	   readed from the standard input 
 *	2) If logfile is specified using both -f switch and without it,
 *	   the -f argument is used and data is readed from that file
 *	3) If logfile is specified but not the keyfile, this will be
 *	   /var/ssyslog/xxx where xxx is the logfile with all '/'
 *	   replaced by '.'
 *
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/syslog.h>
#include <sys/uio.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "hash.h"
#include "../syslogd.h"

char	*corrupted = "corrupted\n";
char	*default_logfile = "/var/log/messages";
char	*keyfile;
char	*key0file;
char	*logfile;
char	*macfile;
int	 method;
int	 use_stdin;

extern char *optarg;


/*
 * release allocated memory
 */
void
release()
{
	if (keyfile != default_keyfile)
		free(keyfile);
	if (key0file)
		free(key0file);
	if (logfile != default_logfile)
		free(logfile);
	if (macfile)
		free(macfile);
}


/*
 * usage
 */
void
usage()
{
	fprintf (stderr,
	"Usage:\n"
	"Check mode:\n"
	"    peochk [-f logfile] [-i key0file] [-k keyfile]"
	"[-m hash_method] [logfile]\n\n"
	"Initial key generator mode:\n"
	"    peochk -g [-k keyfile] [-m hash_method] [logfile]\n\n"
	"hash_method options:\n"
	"\tmd5, rmd160, sha1\n\n"
	"When no logfile is specified or it is without the -f switch "
	"the data is read\n"
	"from the standard input.\n"
 	"If logfile is specified using both -f switch and without "
	"it, the -f argument\n"
	"is used and data is readed from that file.\n"
	"If the logfile is specified but not the keyfile, the default key"
	"file is\n"
	"/var/ssyslog/xxx where xxx is the logfile with all '/' "
	"replaced by '.'\n\n"
	"Default values:\n"
	"\tlogfile    : /var/log/messages\n"
	"\tkeyfile    : /var/ssyslog/.var.log.messages.key\n"
	"\tkey0file   : strcat(keyfile, \"0\");\n"
	"\thash_method: sha1\n\n"
	"\tIf -l switch is specified, the mac'ed log file is "
	"strcat(keyfile, \".mac\");\n");
	exit(1);
}


/*
 * readline
 */
int readline (fd, buf, len)
	int     fd;
	char   *buf;
	size_t  len;
{
	int readed;
	int r;

	readed = 0;
	while (len) {
		if ( (r = read(fd, buf, 1)) == -1)
			return (-1);
		if (!r || *buf == '\n')
			break;
		buf++;
		readed++;
		len--;
	}
	*buf = '\0';
	return readed;
}

		
/*
 * check: read logfile and check it
 */
void
check()
{
	int   i;
	int   input;
	int   mfd;
	char *key;
	int   keylen;
	char  lastkey[41];
	int   lastkeylen;
	int   line;
	char  mkey1[41];
	char  mkey2[41];
	int   mkeylen;
	char  msg[MAXLINE];
	int   msglen;
	
	/* open logfile */
	if (use_stdin)
		input = STDIN_FILENO;
	else if ( (input = open(logfile, O_RDONLY, 0)) == -1)
		err(1, logfile);

	/* open macfile */
	if (macfile)
		if ( (mfd = open(macfile, O_RDONLY, 0)) == -1)
			err(1, macfile);

	/* read initial key */
	if ( (i = open(key0file, O_RDONLY, 0)) == -1)
		err(1, key0file);
	if ( (keylen = read(i, key, 40)) == -1)
		err(1, key0file);
	if (!keylen)
		errx(1, "%s: %s", key0file, corrupted);
	key[keylen] = 0;
	close(i);

	/* read last key */
	if ( (i = open(keyfile, O_RDONLY, 0)) == -1)
		err(1, keyfile);
	if ( (lastkeylen = read(i, lastkey, 40)) == -1)
		err(1, keyfile);
	if (!lastkeylen)
		errx(1, "%s: %s", keyfile, corrupted);
	lastkey[lastkeylen] = 0;
	close(i);

	/* test both key lenghts */
	if (lastkeylen != keylen)
		errx(1, "%s and/or %s %s", key0file, keyfile, corrupted);

	/* check it */
	line = 1;
	msglen = strlen(msg);
	while( (i = readline(input, msg, MAXLINE)) > 0) {
		if (macfile) {
			if ( (mkeylen = mac2(key, keylen, msg, msglen, mkey1)) < 0)
				err(1, macfile);
			if (readline(mfd, mkey2, mkeylen) < 0)
				err(1, macfile);
			if (strncmp(mkey2, mkey1, mkeylen))
				errx(1, "%s %s on line %i\n", logfile, corrupted, line);
			line++;
		}
		if ( (keylen = mac(method, key, keylen, msg, msglen, key)) == -1)
			err(1, "mac function");
	}

	if (macfile)
		close(mfd);

	if (i < 0)
		errx(1, "error reading logs form %s", (use_stdin) ? "standard input" : logfile);

	if (strcmp(lastkey, key)) 
		errx(1, "%s %s\n", logfile, corrupted);

	fprintf(stderr, "%s file is ok\n", logfile);
}


/*
 * generate:
 *	generate initial key and write it on keyfile and key0file
 */
void
generate()
{
	int	 ctimelen;
	char	*ctimestr;
	int	 kfd;
	int	 k0fd;
	char	 newkey[41];
	int	 newkeylen;
	time_t	 t;

	time(&t);
	ctimelen = strlen( (ctimestr = ctime(&t)));
	newkeylen = mac(method, NULL, 0, ctimestr, ctimelen, newkey);
	if ( (kfd = open(keyfile, O_WRONLY|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR)) == -1) {
		release();
		err(1, keyfile);
	}
	if ( (k0fd = open(key0file, O_WRONLY|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR)) == -1) {
		unlink(keyfile);
		close(kfd);
		release();
		err(1, key0file);
	}

	/* write key 0 */
	write(kfd, newkey, newkeylen);
	write(k0fd, newkey, newkeylen);
	close(kfd);
	close(k0fd);
}
	

/*
 * main
 */
int
main (argc, argv)
	int argc;
	char **argv;
{
	int	 action;
	int	 ch;

	/* integrity check mode */
	action = 0;

	/* default values */
	use_stdin = 1;
	logfile = default_logfile;
	keyfile = default_keyfile;
	key0file = NULL;
	macfile = NULL;
	method = SHA1;

	/* parse command line */
	while ( (ch = getopt(argc, argv, "f:ghi:k:lm:")) != -1) {
		switch (ch) {
			case 'f':
				/* log file (intrusion detection mode) */
				if (logfile != default_logfile)
					free(logfile);
				if ( (logfile = strdup(optarg)) == NULL) {
					release();
					err(1, optarg);
				}
				use_stdin = 0;
				break;
			case 'g':
				/* generate new keyfile and initial key */
				action = 1;
				break;
			case 'i':
				/* key 0 file */
				if (key0file)
					free(key0file);
				if ( (key0file = strdup(optarg)) == NULL) {
					release();
					err(1, optarg);
				}
				break;
			case 'k':
				/* keyfile */
				if (keyfile != default_keyfile)
					free(keyfile);
				if ( (keyfile = strdup(optarg)) == NULL) {
					release();
					err(1, optarg); 
				}
				break;
			case 'l':
				macfile = (char*)1;
				break;
			case 'm':
				/* hash method */
				if ( (method = gethash(optarg)) < 0) {
					release();
					usage();
				}
				break;
			case 'h':
			default:
				release();
				usage();
		}

	}

	/* check logfile specified without -f switch */
	argc -= optind;
	argv += optind;
	if (argc && use_stdin)
		if ( (logfile = strdup(argv[argc-1])) == NULL) {
			release();
			err(1, argv[argc-1]);
		}

	/* if keyfile was not specified converted logfile is used instead */
	if (keyfile == default_keyfile && logfile != default_logfile) {
		char *tmp;
		if ( (tmp = strkey(logfile)) == NULL) {
			release();
			err(1, logfile);
		}
		keyfile = (char*) calloc(1, 14+strlen(tmp));
		strcpy(keyfile, "/var/ssyslog/");
		strcat(keyfile, tmp);
		free(tmp);
	}

	/* if key0file was not specified create one */
	if (key0file == NULL) {
		if ( (key0file = calloc(1, strlen(keyfile)+1)) == NULL) {
			release();
			err(1, keyfile);
		}
		strcpy(key0file, keyfile);
		strcat(key0file, "0");
	}

	/* create macfile */
	if (macfile)
		if ( (macfile = strmac(keyfile)) == NULL) {
			release();
			err(1, "creating %s.mac", keyfile);
		}

	/* execute action */
	if (action)
		generate();
	else check();

	release();
	return (0);
}


